README.md
# Accessible Tabbed Interfaces
A script to progressively enhance sectioned content into an accessible tabbed interface.


## How to use 
To help facilitate the simplest integration with your code base, the necessary markup has been boiled down to a wrapping element with a `data-atabs` attribute to serve as the Tab Widget container. Additionally, `tab`s, and `tabpanel`s can be designated via different markup patterns to help suit your needs.  

### Example Setup
```html
<div data-atabs> <!-- necessary wrapping element -->
  <!-- Panel method 1 -->
  <div data-atabs-panel 
    data-atabs-tab-label="Tab label goes here">
    <!-- all panel content goes here -->
  </div>

  <!-- Panel method 2 -->
  <section data-atabs-panel>
    <h# data-atabs-heading>
      <!-- 
        The text/markup injected the panel's 
        associated role="tab" element.
      -->
    </h#>
  </section>
  <!-- repeat as necessary -->
</div>
<!-- ... -->
<script src="index.js"></script>
<script>
  var widget = '[data-atabs]';
  var els = document.querySelectorAll(widget);

  // Generate all Tab Widget instances
  for ( var i = 0; i < els.length; i++ ) {
    var nTabs = new ARIAtabs( els[i] );
  }
</script>
```

### Injecting content into the Tab Widget
Once a new instance of a Tab Widget has been created, the `addTab` function can be called from outside the script.  Using this function, you can reference elements in the DOM (by `id`) that are not in the `data-atabs` wrapping element, and move them into the Tab Widget, creating a `tab` and `tabpanel` for the content. This may be useful if you need to create a Tab Widget, but you may not have full control over the HTML of your document.

For instance:
```html
<script src="index.js"></script>
<script>
  var tabInstance = '[data-atabs]';
  var els = document.querySelectorAll(tabInstance);
  var injectContent = document.getElementById('inject-content');
  var cloneContent = injectContent.cloneNode(true);
  var allTabs = [];

  // Generate all tab instances
  for ( var i = 0; i < els.length; i++ ) {
    var nTabs = new ARIAtabs( els[i] );

    allTabs.push(nTabs);
  }

  // remove the original instance of the external content from the document.
  injectContent.parentNode.removeChild(injectContent);

  // Inject the external content into a particular
  // tab, captured in the allTabs var.
  allTabs[1].addTab(cloneContent, 'Tab label', 'add-a-class');
</script>
```

### Tab Widget attributes & options
The Tab Widget script runs through the markup looking for specific `data-atabs-*` attributes to use as hooks to modify the original markup and generate the final component.  Here are the `data` attributes that are recognized by this script.

* `data-atabs`  
  The primary hook. This necessary attribute is used to contain the final Tab Widget.  
* `data-atabs-panel`    
  Designates that an element should serve as a `tabpanel`. If given the value of "default", the script will set this `tabpanel` and associated `tab` to be active, instead of automatically  first `tab` and `tabpanel`.  If multiple `data-atabs-panel` attributes have the value of "default", only the first one will be respected.
* `data-atabs-tab-label`   
  Also used on the element that will be a `tabpanel`, this attribute indicates that the generated `tab` should use its value as the `tab`'s label. The value of `data-atabs-tab-label` takes precedents over using the content of the `tabpanel`'s heading when generating the `tab`.
* `data-atabs-heading`   
  Place this attribute on the element that serves as the heading within the `tabpanel`. Unless a `data-atabs-tab-label` is used on the `tabpanel`, this heading will serve as the accessible name for the generated `tab`.  By default, elements with the `data-atabs-heading` attribute will be removed after their content has been used for the generated `tab`, unless the value of "keep" is set, e.g. `data-atabs-heading="keep`. Only the first instance of an element with this attribute will be recognized by the script.
* `data-atabs-toc`  
  Without JavaScript, a table of contents (TOC) can provide easy access to different sections of a document that would have otherwise been part of the Tab Widget. With JavaScript available, the TOC isn't as necessary. Providing this attribute with the `id` of the TOC parent element will remove the TOC from the DOM.
* `data-atabs-manual`  
  By default, when keyboard focus is set to a `tab`, its associated `tabpanel` will open by default, and the `tab` will be set to the selected state. 

  If this attribute is set to the `data-atabs` wrapper of *any* Tab Widget in a document, it will make **all** Tab Widgets require manual activation of a `tab` to reveal its associated `tabpanel`.  The reason this globally affects Tab Widgets is to mitigate any possibility of an inconsistent user experience between different Tab Widgets in the same document.  
* `data-atabs-orientation`  
  If this attribute is set to the `data-atabs` wrapper element, and it's value is set to "vertical", then it will add `aria-orientation="vertical"` to the `tablist`. As `aria-orientation` is not well supported across all screen readers, this does not presently have any effect on the current functionality of the Tab Widget.
* `data-atabs-panel-wrap`
  Optional: use this attribute on a wrapping element for all of the `tabpanel`s. Useful for styling a single wrapper for each `tabpanel`.

## User Experience
Tab Widgets are a type of show/hide component, and the manner in which you interact with a Tab Widget will dictate the experience.

For mouse and touch users, the show/hide functionality is initiated by interacting with the `tab`s within the `tablist`.  Activating a `tab` will reveal the `tabpanel` its associated with, while also deactivating the previously active `tab`, and hiding its `tabpanel`.

For keyboard users, and screen reader users in forms mode, a Tab Widget will automatically select and reveal the contents of the focused `tab` as a user navigates the `tablist` with arrow keys.  

A option is available (`data-atabs-manual`) to require manual activation of `tab`s. This is not necessarily preferred behavior of a Tab Widget, but there may be instances where it's required due to performance issues.

The [ARIA Authoring Practices](https://www.w3.org/TR/wai-aria-practices-1.2/#tabpanel) notes:
>It is recommended that tabs activate automatically when they receive focus as long as their associated tab panels are displayed without noticeable latency. This typically requires tab panel content to be preloaded. Otherwise, automatic activation slows focus movement, which significantly hampers users' ability to navigate efficiently across the tab list.  

This script has been tested to work with mouse, touch, and keyboard devices. Each input device has also been re-tested while running various browser and screen reader pairings.

**FYI: More information concerning expected functionality and screen reader announcements will be linked to from here.**

## Dependencies and known issues
There are no dependencies for this script. Any necessary polyfill (for IE11) is included in the JavaScript.  

There are some issues with how screen readers interact with Tab Widgets. Where possible, this script will attempt to mitigate issues with screen readers, but not if it were to create additional bugs if the original but were to be fixed:
* [NVDA: Unexpected focusing of inactive `tab`s in `tablist`](https://github.com/nvaccess/nvda/issues/8906)  
* [JAWS: Unexpected focusing of inactive `tab`s in `tablist`](https://github.com/FreedomScientific/VFO-standards-support/issues/132)
* JAWS + Windows 7 Firefox:  
  Navigating out of a `tablist` by use of <kbd>Tab</kbd> key, focus will be moved to the exposed `tabpanel`. JAWS will not leave forms mode as expected, and will require that the user do so manually to navigate the contents of the `tabpanel`.


## Additional Reading
* [ARIA Specification: Tab Role](https://www.w3.org/TR/wai-aria-1.2/#tab)
* [ARIA Specification: Tablist Role](https://www.w3.org/TR/wai-aria-1.2/#tablist)
* [Aria Specification: Tabpanel Role](https://www.w3.org/TR/wai-aria-1.2/#tabpanel)
* [WAI-ARIA Authoring Practices: Tab Widgets](https://www.w3.org/TR/wai-aria-practices-1.2/#tabpanel)


## License, Thank yous & Such
This script was written by Scott O'Hara: [Website](https://www.scottohara.me), [Twitter](https://twitter.com/scottohara).

It has an [MIT](https://github.com/scottaohara/accessible-components/blob/master/LICENSE.md) license.

Special thanks to [Josh Drumm](https://github.com/wwnjp) and [Chris Ferdinandi](https://gomakethings.com/) for helping me with some JavaScript refactoring and helper functions.

Thanks to [Léonie Watson](https://tink.uk/), Adam Campfield, Ryan Adams, and many others who provided excellent user feedback.

Use it, modify it, contribute to it to help make your project more accessible :)


CHANGELOG.md
# Changelog
All notable changes to this project will be documented in this file.

## [Unreleased]
- [Removable `tab`/`tabpanel` functionality](https://github.com/scottaohara/a11y_tab_widget/issues/6).
- [Open a `tabpanel` from URL hash](https://github.com/scottaohara/a11y_tab_widget/issues/8).
- Additional documentation about UX for Tab Widgets.

## [2.1.4] - 2021-09-20
### Removed
Removed function to re-focus active tab, which was originally put in place to work around buggy JAWS behavior. The function was causing VO focus to constantly re-shift to active tab when attempting to swipe through tabs, due to changes in how VO focus works on iOS.

## [2.1.3] - 2019-08-06
### Added
- Open tabpanel from URL hash via [PR 18](https://github.com/scottaohara/a11y_tab_widget/pull/18).

## [2.1.2] - 2019-01-17
### Added
- New `data-atabs-panel-wrap` attribute allows for a wrapping container for the `tabpanel`s. 

## [2.1.1] - 2018-11-16
### Fixed
- [Home/End keys don't auto activate `tab`s](https://github.com/scottaohara/a11y_tab_widget/issues/14). If statement added to activate a `tab` when navigating with these keys, unless the Tab Widget is set to manual activation.


## [2.1.0] - 2018-11-15
### Fixed
- [Finished version documentation](https://github.com/scottaohara/a11y_tab_widget/issues/10)
- [Support for IE11](https://github.com/scottaohara/a11y_tab_widget/issues/9)
- Missing `role=tabpanel` on panels.  Oops.

### Added
- Updated and added missing documentation for new 2.1 version of this script.
- Various example/test files.
- [Work around for NVDA + JAWS focus bug](https://github.com/scottaohara/a11y_tab_widget/commit/7bda439de03d09c2a472a6ee5d95c2a4b663d679). [NVDA filed bug](https://github.com/nvaccess/nvda/issues/8906), [JAWS filed bug](https://github.com/FreedomScientific/VFO-standards-support).

### Changed
- `tabpanel`s can only receive keyboard focus on initial <kbd>Tab</kbd> away from the `tablist`. Once the `tabpanel` loses focus, the `tabindex` is removed. `tabpanel`s no longer behave as 'clickable' elements when navigating by touch.
- Up, down, left, and right arrow keys all work to navigate between `tab`s within a `tablist`, regardless of orientation. This is due to the fact that `aria-orientation` has poor support and user feedback all requested this feature.
- Change Delete keycode to "delete" and not "Backspace" key (for future release).
- Generate `span`s instead of `button`s.  Add in functionality for <kbd>Space</kbd> and <kbd>Enter</kbd> keys.  This was to attempt to fix strange behavior that didn't actually have anything to do with the `button`s... but since `button`s !== `tab`s, no real reason to change back...
- `aria-controls` is dynamically added to only the active `tab` so that JAWS won't announce key commands to move to an element that is inaccessible, if virtual cursor is on inactive `tab`s.

### Removed
- `:focus-visible` and `.focus-visible` classes. Not really necessary after various updates to the script since the previous release.


## [2.0.1] - 2018-09-20
### Added
- Use `:focus-visible` and `.focus-visible` class in demonstration page. However this polyfill was not added as a script dependency.
### Fixed
- [Make vertical tablists responsive](https://github.com/scottaohara/a11y_tab_widget/issues/7) update to CSS.

## [2.0.0] - 2018-09-19
### Changed
- Rewrote the entire script from scratch into vanilla JavaScript.
- Revised base markup pattern.
- Revised CSS classes.
- Allow for external elements to be added to a Tab Widget, generating both the new `tabpanel` and associated `tab`.
- Options for non-default `aria-orientation=vertial` and automatic activation of `tab`s on focus. 
### Removed
- Non-ARIA Tab Widget version.


## [1.0.0] - 2015-09-30 to [1.1.1] - 2016-12-19
1.0.0 to 1.1.1 were a jQuery based version of Tab Widgets.  
This series of versions allowed for ARIA vs non-ARIA versions of the Tab Widget.

This version of the script also called for a different version of the expected baseline markup, and hooks to make the script work.  




index.js
'use strict';

if ( typeof Object.assign != 'function' ) {
  // Must be writable: true, enumerable: false, configurable: true
  Object.defineProperty(Object, "assign", {
    value: function assign( target, varArgs ) { // .length of function is 2

      if ( target == null ) { // TypeError if undefined or null
        throw new TypeError('Cannot convert undefined or null to object');
      }

      var to = Object(target);

      for ( var index = 1; index < arguments.length; index++ ) {
        var nextSource = arguments[index];

        if ( nextSource != null ) { // Skip over if undefined or null
          for ( var nextKey in nextSource ) {
            // Avoid bugs when hasOwnProperty is shadowed
            if ( Object.prototype.hasOwnProperty.call(nextSource, nextKey) ) {
              to[nextKey] = nextSource[nextKey];
            }
          }
        }
      }
      return to;
    },
    writable: true,
    configurable: true
  });
}

if ( Element && !Element.prototype.matches ) {
  var proto = Element.prototype;
  proto.matches = proto.matchesSelector ||
    proto.mozMatchesSelector || proto.msMatchesSelector ||
    proto.oMatchesSelector || proto.webkitMatchesSelector;
}

// add utilities
var util = {
  keyCodes: {
    UP: 38,
    DOWN: 40,
    LEFT: 37,
    RIGHT: 39,
    HOME: 36,
    END: 35,
    ENTER: 13,
    SPACE: 32,
    DELETE: 46,
    TAB: 9
  },

  generateID: function ( base ) {
    return base + Math.floor(Math.random() * 999);
  },

  getDirectChildren: function ( elm, selector ) {
    return Array.prototype.filter.call(elm.children, function ( child ) {
      return child.matches(selector);
    });
  },

  getUrlHash: function () {
    return window.location.hash.replace('#', '');
  },

  /**
   * Use history.replaceState so clicking through Tabs
   * does not create dozens of new history entries.
   * Browser back should navigate to the previous page
   * regardless of how many Tabs were activated.
   *
   * @param {string} hash
   */
  setUrlHash: function( hash ) {
    if ( history.replaceState ) {
      history.replaceState(null, '', '#' + hash);
    } else {
      location.hash = hash;
    }
  }
};

(function ( w, doc, undefined ) {
  /**
   * ARIA Tabbed Interface
   * Creates a tab list to toggle the visibility of
   * different subsections of a document.
   *
   * Author: Scott O'Hara
   * Version: 2.1.4
   * License: https://github.com/scottaohara/a11y_tab_widget/blob/master/LICENSE
   */
  var ARIAtabsOptions = {
    baseID: 'atab_',
    defaultTabLabel: 'Tab ',
    elClass: 'atabs',
    customTabClassAttribute: 'data-atabs-tab-class',
    tabLabelAttribute: 'data-atabs-tab-label',
    headingAttribute: 'data-atabs-heading',
    defaultOrientation: 'horizontal',
    orientationAttribute: 'data-atabs-orientation',
    panelWrapper: 'data-atabs-panel-wrap',
    panelClass: 'atabs__panel',
    panelSelector: '[data-atabs-panel]',
    tabClass: 'atabs__list__tab',
    tabListClass: 'atabs__list',
    tablistSelector: '[data-atabs-list]',
    manualAttribute: 'data-atabs-manual',
    manual: false
  };


  var ARIAtabs = function ( inst, options ) {
    var _options = Object.assign(ARIAtabsOptions, options);
    var orientation = _options.defaultOrientation;
    var _tabListContainer;
    var _tabs = [];
    var activeIndex = 0;
    var defaultPanel = 0;
    var selectedTab = activeIndex;
    var el = inst;
    var hasPanelWrapper = el.querySelector('[' + _options.panelWrapper + ']');
    var elID;
    var headingSelector = '[' + _options.headingAttribute + ']';

    var init = function () {
      elID = el.id || util.generateID(_options.baseID);

      if ( el.getAttribute(_options.orientationAttribute) === 'vertical' ) {
        orientation = 'vertical';
      }

      if ( el.hasAttribute(_options.manualAttribute) ) {
        _options.manual = true;
      }

      el.classList.add(_options.elClass);

      // find or create the tabList
      _tabListContainer = generateTablistContainer();

      // create the tabs and setup the panels
      buildTabs.call( this );

      // If there's a table of contents for no-js sections,
      // that won't be needed anymore. Remove it.
      deleteTOC();

      if ( activeIndex > -1 ) {
        activateTab();
      }
    }; // init()


    var generateTablistContainer = function () {
      var tabListContainer = el.querySelector(_options.tablistSelector) || doc.createElement('div');
      tabListContainer.setAttribute('role', 'tablist');
      tabListContainer.classList.add(_options.tabListClass);
      tabListContainer.id = elID + '_list';
      tabListContainer.innerHTML = ''; // clear out anything that shouldn't be there
      if ( orientation === 'vertical' ) {
        tabListContainer.setAttribute('aria-orientation', orientation);
      }
      el.insertBefore(tabListContainer, el.querySelector(':first-child'));

      return tabListContainer;
    }; // generateTablistContainer()


    this.addTab = function ( panel, label, customClass ) {
      var customClass = customClass || panel.getAttribute(_options.customTabClassAttribute);

      var generateTab = function ( index, id, tabPanel, customClass ) {
        var newTab = doc.createElement('span');
        newTab.id = elID + '_tab_' + index;
        newTab.tabIndex = -1;
        newTab.setAttribute('role', 'tab');
        newTab.setAttribute('aria-selected', activeIndex === index);
        if ( activeIndex === index ) {
          newTab.setAttribute('aria-controls', id);
        }
        newTab.setAttribute('data-controls', id);
        newTab.innerHTML = tabPanel;
        newTab.classList.add(_options.tabClass);
        if ( customClass ) {
          newTab.classList.add(customClass);
        }

        newTab.addEventListener('click', function () {
          onClick.call( this, index );
          this.focus();
          updateUrlHash();
        }, false);

        newTab.addEventListener('keydown', tabElementPress.bind(this), false);
        // newTab.addEventListener('focus', function () {
        //  checkYoSelf.call( this, index );
        // }, false);

        return newTab;
      };

      var newPanel = panel;
      var i = _tabs.length;

      if ( !newPanel ) {
        return;
      }

      var panelHeading = newPanel.querySelector(headingSelector);
      var finalLabel = [
            label,
            newPanel.getAttribute(_options.tabLabelAttribute),
            panelHeading && panelHeading.textContent,
            _options.defaultTabLabel + (i + 1)
          ]
          .filter( function ( l ) {
            return l && l !== '';
          })[0];

      var newId = newPanel.id || elID + '_panel_' + i;
      var t = generateTab(i, newId, finalLabel, customClass);

      _tabListContainer.appendChild(t);
      newPanel.id = newId;
      newPanel.setAttribute('role', 'tabpanel');
      newPanel.setAttribute('aria-labelledby', elID + '_tab_' + i);
      newPanel.classList.add(_options.panelClass);
      newPanel.hidden = true;

      if ( !el.contains(panel) ) {
        el.appendChild(panel);
      }

      if ( newPanel.getAttribute('id') === util.getUrlHash()) {
        activeIndex = i;
      } else if ( defaultPanel === 0 && newPanel.getAttribute('data-atabs-panel') === 'default' ) {
        activeIndex = i;
        defaultPanel = activeIndex;
      }

      if ( panelHeading ) {
        if ( panelHeading.getAttribute(_options.headingAttribute) !== 'keep' ) {
          panelHeading.parentNode.removeChild(panelHeading);
        }
      }

      newPanel.addEventListener('keydown', panelElementPress.bind(this), false);
      newPanel.addEventListener('blur', removePanelTabindex, false);

      _tabs.push({
        tab: t,
        panel: newPanel
      });
    }; // this.addTab


    var buildTabs = function () {
      var tabs;

      /**
       * Typically tab panels should be direct children of the main tab widget
       * wrapper.  This is necessary so that the script can appropriately associate
       * each tablist with the appropriate tabpanels, allowing for nested tab widgets.
       *
       * If a wrapper for the tabpanels is necessary, for styling/other reasons, then
       * this if statement will look to see if the appropriate panel wrapper div is in
       * place, and if so, adjust the element used to look for the direct children.
       */
      if ( hasPanelWrapper ) {
        tabs = util.getDirectChildren(hasPanelWrapper, _options.panelSelector);
      }
      else {
        tabs = util.getDirectChildren(el, _options.panelSelector);
      }


      for ( var i = 0; i < tabs.length; i++ ) {
        this.addTab(tabs[i]);
      }
    }; // buildTabs()


    var deleteTOC = function () {
      if ( el.getAttribute('data-atabs-toc') ) {
        var toc = doc.getElementById(el.getAttribute('data-atabs-toc'));
        // safety check to make sure a TOC isn't set to be deleted
        // after it's already deleted. e.g. if there are two
        // dat-atabs-toc that equal the same ID.
        if ( toc ) {
          toc.parentNode.removeChild(toc);
        }
      }
    }; // deleteTOC()


    var incrementActiveIndex = function () {
      if ( activeIndex < _tabs.length - 1 ) {
        return ++activeIndex;
      }
      else {
        activeIndex = 0;
        return activeIndex;
      }
    }; // incrementActiveIndex()


    var decrementActiveIndex = function () {
      if ( activeIndex > 0 ) {
        return --activeIndex;
      }
      else {
        activeIndex = _tabs.length - 1;
        return activeIndex;
      }
    }; // decrementActiveIndex()


    var focusActiveTab = function () {
      _tabs[activeIndex].tab.focus();
    }; // focusActiveTab()


    var onClick = function ( index ) {
      activeIndex = index;
      activateTab();
    }; // onClick()


    var moveBack = function ( e ) {
      e.preventDefault();
      decrementActiveIndex();
      focusActiveTab();

      if ( !_options.manual ) {
        activateTab();
        updateUrlHash();
      }
    }; // moveBack()


    var moveNext = function ( e ) {
      e.preventDefault();
      incrementActiveIndex();
      focusActiveTab();

      if ( !_options.manual ) {
        activateTab();
        updateUrlHash();
      }
    }; // moveNext()


    /**
     * A tabpanel is focusable upon hitting the TAB key
     * from a tab within a tablist.  When navigating away
     * from the tabpanel, with the TAB key, remove the
     * tabindex from the tabpanel.
     */
    var panelElementPress = function ( e ) {
      var keyCode = e.keyCode || e.which;

      switch ( keyCode ) {
        case util.keyCodes.TAB:
          removePanelTabindex();
          break;

        default:
          break;
      }
    }; // panelElementPress()


    var removePanelTabindex = function () {
      _tabs[activeIndex].panel.removeAttribute('tabindex');
    }; // removePanelTabindex()


    var tabElementPress = function ( e ) {
      var keyCode = e.keyCode || e.which;

      switch ( keyCode ) {
        case util.keyCodes.TAB:
          _tabs[selectedTab].panel.tabIndex = 0;
          activeIndex = selectedTab;
          break;

        case util.keyCodes.ENTER:
        case util.keyCodes.SPACE:
          e.preventDefault();
          activateTab();
          updateUrlHash();
          break;

        case util.keyCodes.LEFT:
        case util.keyCodes.UP:
          moveBack( e );
          break;

        case util.keyCodes.RIGHT:
        case util.keyCodes.DOWN:
          moveNext( e );
          break;

        case util.keyCodes.END:
          e.preventDefault();
          activeIndex = _tabs.length - 1;
          focusActiveTab();
          if ( !_options.manual ) {
            activateTab();
            updateUrlHash();
          }
          break;

        case util.keyCodes.HOME:
          e.preventDefault();
          activeIndex = 0;
          focusActiveTab();
          if ( !_options.manual ) {
            activateTab();
            updateUrlHash();
          }
          break;

        case util.keyCodes.DELETE:
          if ( _tabs.length > 1 && canRemove ) {
            e.preventDefault();
            removeTabAndPanel(activeIndex);
            focusActiveTab();
          }
          break;

        default:
          break;
      }
    }; // tabElementPress()


    /**
     * This function shouldn't exist.  BUT for...
     * https://github.com/nvaccess/nvda/issues/8906
     * https://github.com/FreedomScientific/VFO-standards-support/issues/132
     *
     * Note this doesn't completely fix JAWS announcements.
     * With this function, focus will be placed on the correct Tab,
     * but JAWS will make no announcement until the user begins
     * re-navigating with arrow keys.
     *
     * The alternative is not using this, having JAWS announce the
     * inactive tag (which will receive focus), JAWS will announce
     * to use the Space key to activate, but nothing will happen.
     */
    // sept19-2021 - commenting this out as it causes focus issues with 
    // iOS + VoiceOver.  
    // var checkYoSelf = function ( index ) {
    //  if ( index !== activeIndex ) {
    //    focusActiveTab();
    //  }
    // }; // checkYoSelf()


    var deactivateTabs = function () {
      for ( var i = 0; i < _tabs.length; i++ ) {
        deactivateTab(i);
      }
    }; // deactivateTabs()


    var deactivateTab = function ( idx ) {
      _tabs[idx].panel.hidden = true;
      _tabs[idx].tab.tabIndex = -1;
      _tabs[idx].tab.setAttribute('aria-selected', false);
      _tabs[idx].tab.removeAttribute('aria-controls');
      // remove the aria-controls from inactive tabs since
      // a user can *not* move to their associated element
      // if that element is not displayed.
    }; // deactivateTab()


    /**
     * Update the active Tab and make it focusable.
     * Deactivate any previously active Tab.
     * Reveal active Panel.
     */
    var activateTab = function () {
      var active = _tabs[activeIndex];
      deactivateTabs();
      active.tab.setAttribute('aria-controls', active.tab.getAttribute('data-controls'));
      active.tab.setAttribute('aria-selected', true);
      active.tab.tabIndex = 0;
      active.panel.hidden = false;
      selectedTab = activeIndex;
      return selectedTab;
    }; // activateTab()


    /**
     * Update URL Hash so direct link to the currently open Tab is exposed for copy & paste.
     */
    var updateUrlHash = function () {
      var active = _tabs[activeIndex];
      util.setUrlHash(active.tab.getAttribute('data-controls'));
    }; // updateUrlHash()


    init.call( this );
    return this;
  }; // ARIAtabs()


  w.ARIAtabs = ARIAtabs;
})( window, document );
